common/persistence/elasticsearch/esProcessor.go

```diff
@@ -197,42 +196,79 @@ func (p *esProcessorImpl) bulkAfterAction(_ int64, requests []elastic.BulkableRe
 	if err != nil {
 		// This happens after configured retry, which means something bad happens on cluster or index
 		// When cluster back to live, processor will re-commit those failure requests
-		p.logger.Error("Error commit bulk request.", tag.Error(err))
 
+		isRetryable := isRetryableError(err)
+		p.logger.Error("Unable to commit bulk ES request.", tag.Error(err), tag.Bool(isRetryable))
 		for _, request := range requests {
 			p.logger.Error("ES request failed.", tag.ESRequest(request.String()))
 			p.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorFailures)
+
+			if !isRetryable {
+				visibilityTaskKey := p.extractVisibilityTaskKey(request)
+				if visibilityTaskKey == "" {
+					continue
+				}
+				p.sendToAckChan(visibilityTaskKey, false)
+			}
 		}
 		return
 	}
 
+	responseIndex := p.buildResponseIndex(response)
 	for i, request := range requests {
-		visibilityTaskKey := p.getVisibilityTaskKey(request)
+		visibilityTaskKey := p.extractVisibilityTaskKey(request)
 		if visibilityTaskKey == "" {
 			continue
 		}
-		responseItem := response.Items[i]
-		for _, resp := range responseItem {
+
+		docID := p.extractDocID(request)
+		if docID == "" {
+			p.sendToAckChan(visibilityTaskKey, false)
+			continue
+		}
+
+		responseItems, ok := responseIndex[docID]
+		if !ok {
+			p.logger.Error("ES request failed. Request item doesn't have corresponding response item.",
+				tag.Value(i),
+				tag.Key(visibilityTaskKey),
+				tag.ESDocID(docID),
+				tag.ESRequest(request.String()))
+			p.sendToAckChan(visibilityTaskKey, false)
+			continue
+		}
+
+		for _, responseItem := range responseItems {
 			switch {
-			case isResponseSuccess(resp.Status):
+			case isSuccessStatus(responseItem.Status):
 				p.sendToAckChan(visibilityTaskKey, true)
-			case !isResponseRetryable(resp.Status):
+			case !isRetryableStatus(responseItem.Status):
 				p.logger.Error("ES request failed.",
```

Comment:
unrelated question, but if we attempt to delete the first time, and it succeeds, but we get an error, and we retry and get a 404, do we count that as a failure? or does deleting a non-existent item return success?

Reply:
ES returns 404 for deleting of non-existing item and we will count it as failure. I believe it is correct behaviour.